compat/mingw.c

```diff
@@ -3587,8 +3587,29 @@ static void setup_windows_environment(void)
 			struct strbuf buf = STRBUF_INIT;
 			strbuf_addstr(&buf, tmp);
 			if ((tmp = getenv("HOMEPATH"))) {
+				/*
+				 * Since Vista, HOMEDRIVE/HOMEPATH are set to Windows\System32
+				 * for secondary logons/services instead of being unset.
+				 * We don't want to (and likely can't) save stuff there
+				 * so ignore them in such a case.
+				 */
+				wchar_t w_system_dir[MAX_PATH];
+				wchar_t *w_buf;
+				size_t len_w_buf;
 				strbuf_addstr(&buf, tmp);
-				if (is_directory(buf.buf))
+				if (!(GetSystemDirectoryW(
+						(LPWSTR)w_system_dir,
+						ARRAY_SIZE(w_system_dir))))
+					die("GetSystemDirectoryW failed: %ld",
+					    GetLastError());
+				len_w_buf = buf.len + 1;
+				w_buf = alloca(len_w_buf * sizeof(wchar_t));
```

Comment:
In the rest of `compat/mingw.c`, we do not use `alloca()` for buffers that receive the output of `xutftowcs()`, but we use stack allocation directly (`wchar_t wpathname[MAX_LONG_PATH];`).

Besides, it strikes me as the wrong way round to convert the UTF-8 encoded candidate to UTF-16 to compare it to the system32 directory's path. In Git for Windows' source code, we usually handle paths in UTF-8, therefore the output of `GetSystemDirectoryW()` should be converted to UTF-8 instead, and _then_ we compare.

In this instance, I further would strongly suggest to untangle the post-image of this diff by extracting the code to determine the path of the system32 directory to its own function. Something like this:

```c
static const char *get_system32_path(void)
{
        static char path[MAX_LONG_PATH];
        static int initialized;

        if (!initialized) {
                WCHAR wpath[MAX_LONG_PATH];

                if (!GetSystemDirectoryW(wpath, ARRAY_SIZE(wpath)) ||
                    xwcstoutf(path, wpath, sizeof(path) < 0)
                        /* Ignore failures silently */
                        path[0] = '\0';
                initialized = 1;
        }

        return path;
}
```

Reply:
It's a valid strategy to use `alloca()`. It's just not a strategy that we commonly use in Git's source code. And since it makes it harder to maintain inconsistent code, I would insist on imitating `compat/mingw.c`'s existing code style better.